File: android/src/main/kotlin/io/getimpulse/player/plugin/core/PluginMethod.kt
package io.getimpulse.player.plugin.core

import android.content.Context
import android.graphics.Typeface
import android.graphics.fonts.FontVariationAxis
import io.flutter.plugin.common.MethodCall
import io.getimpulse.player.ImpulsePlayer
import io.getimpulse.player.model.ImpulsePlayerAppearance
import io.getimpulse.player.model.ImpulsePlayerFont
import io.getimpulse.player.model.ImpulsePlayerSettings
import kotlin.math.roundToInt

internal sealed class PluginMethod {

    sealed class Result {
        object Executed : Result()
        data class Data<T>(val value: T) : Result()
    }

    data class Load(val id: Int, val title: String?, val subtitle: String?, val url: String) :
        PluginMethod()

    data class Play(val id: Int) : PluginMethod()
    data class Pause(val id: Int) : PluginMethod()
    data class Seek(val id: Int, val time: Long) : PluginMethod()
    data class GetState(val id: Int) : PluginMethod()
    data class IsPlaying(val id: Int) : PluginMethod()
    data class GetProgress(val id: Int) : PluginMethod()
    data class GetDuration(val id: Int) : PluginMethod()
    data class GetError(val id: Int) : PluginMethod()
    data class SetAppearance(
        val h3: ImpulsePlayerFont,
        val h4: ImpulsePlayerFont,
        val s1: ImpulsePlayerFont,
        val l4: ImpulsePlayerFont,
        val l7: ImpulsePlayerFont,
        val p1: ImpulsePlayerFont,
        val p2: ImpulsePlayerFont,
        val accentColor: Int,
    ) : PluginMethod()
    data class SetSettings(
        val pictureInPictureEnabled: Boolean,
        val castReceiverApplicationId: String?,
    ) : PluginMethod()

    fun execute(): Result = when (this) {
        is Load -> {
            PluginNativeViewFactory.get(id)?.load(title, subtitle, url)
            Result.Executed
        }

        is Play -> {
            PluginNativeViewFactory.get(id)?.play()
            Result.Executed
        }

        is Pause -> {
            PluginNativeViewFactory.get(id)?.pause()
            Result.Executed
        }

        is Seek -> {
            PluginNativeViewFactory.get(id)?.seek(time)
            Result.Executed
        }

        is GetState -> {
            val state = PluginNativeViewFactory.get(id)?.getState()
            Result.Data(state)
        }

        is IsPlaying -> {
            val playing = PluginNativeViewFactory.get(id)?.isPlaying()
            Result.Data(playing)
        }

        is GetProgress -> {
            val progress = PluginNativeViewFactory.get(id)?.getProgress()
            Result.Data(progress)
        }

        is GetDuration -> {
            val duration = PluginNativeViewFactory.get(id)?.getDuration()
            Result.Data(duration)
        }

        is GetError -> {
            val error = PluginNativeViewFactory.get(id)?.getError()
            Result.Data(error)
        }

        is SetAppearance -> {
            ImpulsePlayer.setAppearance(
                ImpulsePlayerAppearance(
                    h3, h4, s1, l4, l7, p1, p2,
                    accentColor,
                )
            )
            Result.Executed
        }

        is SetSettings -> {
            ImpulsePlayer.setSettings(
                ImpulsePlayerSettings(
                    pictureInPictureEnabled,
                    castReceiverApplicationId,
                )
            )
            Result.Executed
        }
    }

    companion object {
        fun from(call: MethodCall, context: Context?): PluginMethod? {
            val id = call.argument<Int>(PluginConstants.Parameter.Id)
            return when (call.method) {
                PluginConstants.Method.Load -> {
                    requireNotNull(id)
                    val url = call.argument<String>(PluginConstants.Parameter.Url)
                    val title = call.argument<String>(PluginConstants.Parameter.Title)
                    val description = call.argument<String>(PluginConstants.Parameter.Subtitle)
                    requireNotNull(url)
                    Load(id, title, description, url)
                }

                PluginConstants.Method.Play -> {
                    requireNotNull(id)
                    Play(id)
                }

                PluginConstants.Method.Pause -> {
                    requireNotNull(id)
                    Pause(id)
                }

                PluginConstants.Method.Seek -> {
                    requireNotNull(id)
                    val time = call.argument<Long>(PluginConstants.Parameter.Time)
                    requireNotNull(time)
                    Seek(id, time)
                }

                PluginConstants.Method.GetState -> {
                    requireNotNull(id)
                    GetState(id)
                }

                PluginConstants.Method.IsPlaying -> {
                    requireNotNull(id)
                    IsPlaying(id)
                }

                PluginConstants.Method.GetProgress -> {
                    requireNotNull(id)
                    GetProgress(id)
                }

                PluginConstants.Method.GetDuration -> {
                    requireNotNull(id)
                    GetDuration(id)
                }

                PluginConstants.Method.GetError -> {
                    requireNotNull(id)
                    GetError(id)
                }

                PluginConstants.Method.SetAppearance -> {
                    requireNotNull(context)
                    SetAppearance(
                        createFont(context, call.argument<Map<String, Any>>("h3")!!),
                        createFont(context, call.argument<Map<String, Any>>("h4")!!),
                        createFont(context, call.argument<Map<String, Any>>("s1")!!),
                        createFont(context, call.argument<Map<String, Any>>("l4")!!),
                        createFont(context, call.argument<Map<String, Any>>("l7")!!),
                        createFont(context, call.argument<Map<String, Any>>("p1")!!),
                        createFont(context, call.argument<Map<String, Any>>("p2")!!),
                        call.argument<Int>("accent_color")!!,
                    )
                }

                PluginConstants.Method.SetSettings -> {
                    val pictureInPictureEnabled = call.argument<Boolean>(PluginConstants.Parameter.PictureInPictureEnabled)
                    val castReceiverApplicationId = call.argument<String>(PluginConstants.Parameter.CastReceiverApplicationId)
                    requireNotNull(pictureInPictureEnabled)
                    SetSettings(
                        pictureInPictureEnabled,
                        castReceiverApplicationId,
                    )
                }

                else -> null
            }
        }

        private fun createFont(context: Context, json: Map<String, Any>): ImpulsePlayerFont {
            val size = (json["size"] as Double).roundToInt()
            val family = json["family"] as String
            val weight = json["weight"] as Int
            val style = json["style"] as String
            val path = when (style) {
                PluginConstants.Value.Normal -> "fonts/$family.ttf"
                PluginConstants.Value.Italic -> "fonts/$family-$style.ttf"
                else -> throw IllegalArgumentException("Unhandled style $style")
            }
            val variationWeight = FontVariationAxis("wght", weight.toFloat())
            return ImpulsePlayerFont(
                size,
                Typeface.Builder(context.assets, path)
                    .setFontVariationSettings(arrayOf(variationWeight))
                    .build()
            )
        }
    }
}

File: ios/Classes/Core/PluginMethod.swift
import UIKit
import Flutter
import ImpulsePlayer

enum PluginMethod {
    
    enum Result<T> {
        case executed
        case data(value: T)
    }
    
    case setCastEnabled(id: Int64, enabled: Bool)
    case load(id: Int64, url: String, title: String?, subtitle: String?, headers: [String:String])
    case play(id: Int64)
    case pause(id: Int64)
    case seek(id: Int64, time: Int64)
    case getState(id: Int64)
    case isPlaying(id: Int64)
    case getProgress(id: Int64)
    case getDuration(id: Int64)
    case getError(id: Int64)
    case setAppearance(h3: ImpulsePlayerFont, h4: ImpulsePlayerFont, s1: ImpulsePlayerFont, l4: ImpulsePlayerFont, l7: ImpulsePlayerFont, p1: ImpulsePlayerFont, p2: ImpulsePlayerFont, accentColor: UIColor)
    case setSettings(pictureInPictureEnabled: Bool, castReceiverApplicationId: String?)

    @MainActor func execute() -> Any? {
        switch self {
        case .setCastEnabled(let id, let enabled):
            PluginNativeViewFactory.get(id)?.setCastEnabled(enabled)
            return nil
            
        case .load(let id, let url, let title, let subtitle, let headers):
            PluginNativeViewFactory.get(id)?.load(url: url, title: title, subtitle: subtitle, headers: headers)
            return nil
            
        case .play(let id):
            PluginNativeViewFactory.get(id)?.play()
            return nil
            
        case .pause(let id):
            PluginNativeViewFactory.get(id)?.pause()
            return nil
            
        case .seek(let id, let time):
            PluginNativeViewFactory.get(id)?.seek(to: time)
            return nil
            
        case .getState(let id):
            let state = PluginNativeViewFactory.get(id)?.getState()
            return state
            
        case .isPlaying(let id):
            let playing = PluginNativeViewFactory.get(id)?.isPlaying()
            return playing
            
        case .getProgress(let id):
            let progress = PluginNativeViewFactory.get(id)?.getProgress()
            return progress
            
        case .getDuration(let id):
            let duration = PluginNativeViewFactory.get(id)?.getDuration()
            return duration
            
        case .getError(let id):
            let error = PluginNativeViewFactory.get(id)?.getError()
            return error
            
        case .setAppearance(let h3, let h4, let s1, let l4, let l7, let p1, let p2, let accentColor):
            ImpulsePlayer.setAppearance(appearance: ImpulsePlayerAppearance(
                h3: h3,
                h4: h4,
                s1: s1,
                l4: l4,
                l7: l7,
                p1: p1,
                p2: p2,
                accentColor: accentColor
            ))
            return nil
            
        case .setSettings(let pictureInPictureEnabled, let castReceiverApplicationId):
            let settings = ImpulsePlayerSettings(
                pictureInPictureEnabled: pictureInPictureEnabled,
                castReceiverApplicationId: castReceiverApplicationId
            )
            ImpulsePlayer.setSettings(settings: settings)
            return nil
        }
    }
    
    static func from(call: FlutterMethodCall) -> PluginMethod? {
        guard let arguments = call.arguments as? Dictionary<String, Any> else { return nil }
        
        let id = arguments[PluginConstants.Parameter.id] as? Int64
        switch call.method {
        case PluginConstants.Method.setCastEnabled:
            guard let id else { fatalError("Missing id") }
            guard let enabled = arguments[PluginConstants.Parameter.enabled] as? Bool else {
                return nil
            }
            return .setCastEnabled(id: id, enabled: enabled)
            
        case PluginConstants.Method.load:
            guard let id,
                  let url = arguments[PluginConstants.Parameter.url] as? String
            else {
                return nil
            }
            let title = arguments[PluginConstants.Parameter.title] as? String
            let subtitle = arguments[PluginConstants.Parameter.subtitle] as? String
            let headers = arguments[PluginConstants.Parameter.headers] as? [String: String] ?? [:]
            return .load(id: id, url: url, title: title, subtitle: subtitle, headers: headers)
            
        case PluginConstants.Method.play:
            guard let id else { fatalError("Missing id") }
            return .play(id: id)
            
        case PluginConstants.Method.pause:
            guard let id else { fatalError("Missing id") }
            return .pause(id: id)
            
        case PluginConstants.Method.seek:
            guard let id else { fatalError("Missing id") }
            guard let time = arguments[PluginConstants.Parameter.time] as? Int64 else {
                return nil
            }
            return .seek(id: id, time: time)
            
        case PluginConstants.Method.getState:
            guard let id else { fatalError("Missing id") }
            return .getState(id: id)
            
        case PluginConstants.Method.isPlaying:
            guard let id else { fatalError("Missing id") }
            return .isPlaying(id: id)
            
        case PluginConstants.Method.getProgress:
            guard let id else { fatalError("Missing id") }
            return .getProgress(id: id)
            
        case PluginConstants.Method.getDuration:
            guard let id else { fatalError("Missing id") }
            return .getDuration(id: id)
            
        case PluginConstants.Method.getError:
            guard let id else { fatalError("Missing id") }
            return .getError(id: id)
            
        case PluginConstants.Method.setAppearance:
            return .setAppearance(
                h3: createFont(json: arguments["h3"] as! [String: Any]),
                h4: createFont(json: arguments["h4"] as! [String: Any]),
                s1: createFont(json: arguments["s1"] as! [String: Any]),
                l4: createFont(json: arguments["l4"] as! [String: Any]),
                l7: createFont(json: arguments["l7"] as! [String: Any]),
                p1: createFont(json: arguments["p1"] as! [String: Any]),
                p2: createFont(json: arguments["p2"] as! [String: Any]),
                accentColor: colorFromInt64(argb: arguments["accent_color"] as! Int64)
            )
            
        case PluginConstants.Method.setSettings:
            guard let pictureInPictureEnabled = arguments[PluginConstants.Parameter.pictureInPictureEnabled] as? Bool else {
                return nil
            }
            let castReceiverApplicationId = arguments[PluginConstants.Parameter.castReceiverApplicationId] as? String
            return .setSettings(
                pictureInPictureEnabled: pictureInPictureEnabled,
                castReceiverApplicationId: castReceiverApplicationId
            )
            
        default:
            return nil
        }
    }
    
    private static func createFont(json: [String: Any]) -> ImpulsePlayerFont {
        let size = json["size"] as! CGFloat
        let family = json["family"] as! String
        let weight = json["weight"] as! CGFloat
        let style = json["style"] as! String
        let bold = weight >= 500
        let italic = style == PluginConstants.Value.italic
        return ImpulsePlayerFont(
            fontType: .customByFamily(familyName: family, bold: bold, italic: italic),
            size: size
        )
    }
    
    private static func colorFromInt64(argb: Int64) -> UIColor {
        let alpha = CGFloat((argb >> 24) & 0xFF) / 255.0
        let red = CGFloat((argb >> 16) & 0xFF) / 255.0
        let green = CGFloat((argb >> 8) & 0xFF) / 255.0
        let blue = CGFloat(argb & 0xFF) / 255.0
        
        return UIColor(red: red, green: green, blue: blue, alpha: alpha)
    }
}
